### PR TITLE
feat(turborepo): port git-based hashing with inputs

### PR DESCRIPTION
**Pull request by gsoltis**
### Description

 - ports getPackageFileHashesFromInputs to rust
 - ports the remainder of the test that deals with inputs from `package_deps_hash_test.go`
 - adds a `relative_path_to` fn to `AbsoluteSystemPath` for getting relative paths that allow directory traversal

### Testing Instructions

 - Existing test suite
 - Ported inputs test
 - New test for `relative_path_to`

### CODE REVIEW

File: crates/turborepo-paths/src/absolute_system_path.rs [hunk 1]
```diff
@@ -197,6 +197,45 @@ impl AbsoluteSystemPath {
     pub fn components(&self) -> Components<'_> {
         self.0.components()
     }
+
+    // Produces a path from self to other, which may include directory traversal
+    // tokens. Given that both parameters are absolute, we _should_ always be
+    // able to produce such a path. The exception is when crossing drive letters
+    // on Windows, where no such path is possible. Since a repository is
+    // expected to only reside on a single drive, this shouldn't be an issue.
+    pub fn relative_path_to(&self, other: &Self) -> AnchoredSystemPathBuf {
```
Comment (NicholasLYang) [inline crates/turborepo-paths/src/absolute_system_path.rs:206]: If `self` and `other` are the same path, we'd end up with an empty path, will that work okay?

**Review by NicholasLYang — APPROVED** [review]

File: crates/turborepo-scm/src/hash_object.rs [hunk 2]
```diff
@@ -71,7 +74,10 @@ fn read_object_hashes<R: Read, W: Write + Send>(
             reader.read_exact(&mut buffer)?;
             let hash = parse_hash_object(&buffer)?;
             let hash = String::from_utf8(hash.to_vec())?;
-            let path = filename.strip_prefix(pkg_prefix)?;
+            // TODO: filename.make_relative_to(pkg_prefix)
+            let full_file_path = git_prefix.join_unix_path(filename)?;
+            let path = pkg_path.relative_path_to(&full_file_path).to_unix()?;
+            //let path = filename.strip_prefix(pkg_prefix)?;
```
Comment (chris-olszewski) [inline crates/turborepo-scm/src/hash_object.rs:80]: I think these comments can get removed?
```suggestion
            let full_file_path = git_prefix.join_unix_path(filename)?;
            let path = pkg_path.relative_path_to(&full_file_path).to_unix()?;
```

**Review by chris-olszewski — APPROVED** [review]

File: crates/turborepo-ffi/src/lib.rs [hunk 3]
```diff
@@ -381,6 +381,91 @@ pub extern "C" fn get_package_file_hashes_from_processing_git_ignore(buffer: Buf
     response.into()
 }
 
+#[no_mangle]
+pub extern "C" fn get_package_file_hashes_from_inputs(buffer: Buffer) -> Buffer {
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-ffi/src/lib.rs:385]: Yeah if we were investing in the sandwich long term I'd definitely push the pattern of a wrapper that returns `Result`. And probably some macros for unwrapping everything cleanly.

File: crates/turborepo-globwalk/src/lib.rs [hunk 4]
```diff
@@ -244,8 +244,8 @@ fn collapse_path(path: &str) -> Option<(Cow<str>, usize)> {
     for segment in path.trim_start_matches('/').split('/') {
         match segment {
             ".." => {
-                lowest_index.get_or_insert(stack.len());
                 stack.pop()?;
+                lowest_index.get_or_insert(stack.len());
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-globwalk/src/lib.rs:248]: Heh, so I was trying to reproduce the error that lead to this change, and I couldn't. After digging a little more, I determined that the position of the line is actually irrelevant, even if the new position is technically more correct. The comparison below always ends up setting the value to the size of the now-popped stack, so a test doesn't catch it if it's in the wrong spot.

I am suspicious of what inputs I had originally sent in that caused an error, but all of the inputs I was using are captured in the inputs tests. Maybe there was an upstream issue that I had resolved? But puzzling...

I'll add a comment.

File: crates/turborepo-ffi/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-ffi/src/lib.rs:385]: I think I'd prefer to keep everything function-scoped in this particular file. It'll make it easier to remove calls, and some of that is coming in the next PR

File: crates/turborepo-scm/Cargo.toml [hunk 5]
```diff
@@ -9,8 +9,10 @@ license = "MPL-2.0"
 [dependencies]
 bstr = "1.4.0"
 git2 = { version = "0.16.1", default-features = false }
+globwalk = { version = "0.1.0", path = "../turborepo-globwalk" }
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-scm/Cargo.toml:12]: I shamelessly copied this line from another `Cargo.toml`. But no, it doesn't make sense to set both. Removed `version`

File: crates/turborepo-scm/src/package_deps.rs [hunk 6]
```diff
@@ -19,12 +22,69 @@ pub fn get_package_file_hashes_from_git_index(
...
+) -> Result<GitHashes, Error> {
+    // TODO: memoize git root -> turbo root calculation once we aren't crossing ffi
+    let git_root = find_git_root(turbo_root)?;
+    let full_pkg_path = turbo_root.resolve(package_path);
+    let package_unix_path_buf = package_path.to_unix()?;
+    let package_unix_path = package_unix_path_buf.as_str()?;
+
+    let mut inputs = inputs
+        .iter()
+        .map(|s| s.as_ref().to_string())
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-scm/src/package_deps.rs:42]: There might be a way to do it, but it wasn't immediately clear to me. The issue I ran into was that we are iterating over a slice, and so can't take ownership over the elements, which is required for using `Into`.

File: crates/turborepo-paths/src/absolute_system_path.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsoltis) [reply crates/turborepo-paths/src/absolute_system_path.rs]: Good suggestion

**Pull request merged**